### PR TITLE
add postgre backend for vault

### PR DESCRIPTION
**Pull request by huskercane**
We use postgre for db

this is the code that I need to store information in postgre db

### CODE REVIEW

File: physical/postgresql.go [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+	defer b.l.Unlock()
+
+	db, err := sql.Open("postgres", b.Url)
+	if err != nil {
+		Trace.Println("< List error db")
+		Error.Fatal(err)
+	}
+	defer db.Close()
+
+	query := "%" + prefix + "%"
```
Comment (sethvargo) [inline physical/postgresql.go]: This looks like an invalid unicode character?

File: physical/postgresql.go [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"sync"
+	"time"
+)
+
+// PostGresBackend is a physical backend that stores data on postgres database
+type PostGresBackend struct {
```
Comment (sethvargo) [inline physical/postgresql.go]: I think we might want `PostgresBackend` instead of `PostGresBackend`

File: physical/physical.go [hunk 3]
```diff
@@ -85,4 +85,5 @@ var BuiltinBackends = map[string]Factory{
 	"s3":        newS3Backend,
 	"etcd":      newEtcdBackend,
 	"mysql":     newMySQLBackend,
+	"postgres":  newPostGresBackend,
```
Comment (armon) [inline physical/physical.go]: Can we change this to "postgresql"

File: physical/postgresql.go [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+	Url string
+
+	l sync.Mutex
+}
+
+var (
```
Comment (armon) [inline physical/postgresql.go]: We do not want these loggers to be static globals

File: physical/postgresql.go [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+	Trace   *log.Logger
+	Info    *log.Logger
+	Warning *log.Logger
+	Error   *log.Logger
+)
+
+// newFileBackend constructs a Filebackend using the given directory
+func newPostGresBackend(conf map[string]string) (Backend, error) {
```
Comment (armon) [inline physical/postgresql.go]: Should be `newPostgresqlBackend`, also the doc string is not correct

**Comment by huskercane** [comment]
Thank you for the comments. 

let me try again :-)

File: physical/postgresql.go [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+	if err != nil {
+		Error.Printf("%q: %s\n", err, sqlStmt)
+		Error.Println("< newPostGresBackend(err)")
+		Error.Panic(err)
+	}
+
+	sqlStmt = `
+CREATE TABLE if not exists vault (
+  id UUID PRIMARY KEY NOT NULL DEFAULT uuid_generate_v4(),
+  key TEXT not null,
```
Comment (BRMatt) [inline physical/postgresql.go]: I noticed the MySQL backend [uses the `key` as the primary key](https://github.com/hashicorp/vault/blob/master/physical/mysql.go#L84). Is it worth doing the same here, or at least adding a uniqueness constraint on the `key` field? [This thread](http://www.postgresql.org/message-id/1185022160.688503.236330@n2g2000hse.googlegroups.com) seems to imply that the cost of maintaining two indexes may not be worthwhile, especially if the id index is never used.

File: physical/postgresql.go [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+		}
+
+	} else {
+		stmt, err := txn.Prepare("insert into vault (key, value, created_at, updated_at) values ($1, $2, $3, $4)")
+		if err != nil {
+			Trace.Println("< Put error insert prepare")
+			Error.Fatal(err)
+		}
+
+		time := time.Now()
```
Comment (BRMatt) [inline physical/postgresql.go]: It might be worth using According to the [package docs](http://golang.org/pkg/time/#Now), `Now()` returns the machine's local time. Given that the audit log [uses UTC](https://github.com/hashicorp/vault/blob/e2e077964b5200208a7554963d49f55ec98ade57/audit/format_json.go#L33) it might be worthing using it here too. It seems it's even possible to have postgres [use the current UTC time for the column's default value](http://stackoverflow.com/a/16610360/149297).

File: physical/postgresql.go [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+	Trace.Println("< newPostGresBackend")
+	return &PostGresBackend{Url: url}, nil
+}
+
+func (b *PostGresBackend) Delete(k string) error {
+	Trace.Println("> Delete")
+	b.l.Lock()
+	defer b.l.Unlock()
+
+	db, err := sql.Open("postgres", b.Url)
```
Comment (BRMatt) [inline physical/postgresql.go]: According to the golang documentation the `DB` value returned by `Open()` sets up a connection pool that is [safe for concurrent use by multiple goroutines](http://golang.org/pkg/database/sql/#DB), so it may be worth opening the connection when the backend is initialized, then changing these functions to use the pool.

File: physical/postgresql.go [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+	stmt, err := txn.Prepare("delete from vault where key = $1")
+	if err != nil {
+		Trace.Println("< Delete error prepare")
+		Error.Fatal(err)
+	}
+
+	_, err = stmt.Exec(k)
+	if err != nil {
+		Trace.Println("< Delete error exec")
+		Error.Fatal(err)
```
Comment (BRMatt) [inline physical/postgresql.go]: I'm not sure if the postgres backend should panic and crash the vault instance if it can't run a query! :smile:  Usually go programs tend to return errors the encounter to the caller. Sometimes they [wrap them in another error](https://github.com/hashicorp/vault/blob/e2e077964b5200208a7554963d49f55ec98ade57/physical/consul.go#L103) to give some context as to what area generated the error.

**Comment by johnjelinek** [comment]
:+1:

**Comment by huskercane** [comment]
Hello

Can you please review code again?

File: physical/postgresql.go [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+		Value: result,
+	}
+
+	return ent, nil
+}
+
+// Put - Update a row in database
+func (b *PostgresqlBackend) Put(entry *Entry) error {
+
+	// TODO: fix me
```
Comment (jefferai) [inline physical/postgresql.go]: Does this need to be fixed?  :-)

**Comment by huskercane** [comment]
no code is fine.

just errant comment

**Comment by jefferai** [comment]
Closing in favor of #945
